### ISSUE TITLE
Add ability to mock WebSocket connections in Cypress

### ISSUE THREAD

**Issue by 0xApotheosis** [shapeshift/web#949]
# **IMPORTANT: NO BOUNTIES WILL BE PAID FROM 2/5 - 2/14 - IF YOU COMPLETE THIS BOUNTY DURING THIS TIMEFRAME WE WILL REMIT PAYMENT ON 2/14**

# Background 

[Cypress](https://docs.cypress.io/guides/overview/why-cypress) supports [intercepting and mocking REST communications](https://docs.cypress.io/guides/guides/network-requests) out of the box. It does not, however, support WebSocket connections in the same way.

We use WebSockets to provide data to the web Dashboard, notably transaction data relating to the connected wallet. We are not currently able to test the rendering of transactions due to this Cypress limitation.

# Other considerations

A number of libraries exist that may be of use, e.g.:

- [cypress-websocket-testing](https://github.com/lensesio/cypress-websocket-testing)
- [mock-socket](https://github.com/thoov/mock-socket)
- ...

An example response from a WebSocket server relating to transactions is available in `cypress/factories/ethereum/transactions.ts`

We have several intercepted REST endpoints [here](https://github.com/shapeshift/web/blob/08f03825e01401033b94da0903713c8c39ec9229/cypress/support/commands.ts#L112) that could be used for inspiration - we probably want to have a Cypress WebSocket command implemented similarly.

# Acceptance Criteria

- Add the ability to mock WebSocket connections for Cypress tests
- Implement an example of a mock WebSocket server for transactions, with an appropriate test to match (e.g. testing that the mocked transactions render on the Dashboard correctly in `TransactionRow` when used in `RecentTransactions`, `TxHistory`, etc). This test will likely sit in `cypress/integration/dashboard_spec.ts`, though could be broken out if the tests are substantial enough.
- The implementation should be extensible and show us the best practice for interacting with WebSockets from Cypress moving forward

# Bounty Documentation 
- Please check out [our docs](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties) for how to get in touch with us or ask questions.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 2200.0 FOX (799.36 USD @ $0.36/FOX)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/shapeshift/web/949/100027615).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/shapeshift/web/949/100027615).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://discord.gg/gitcoin/'>Gitcoin's Discord</a>
* $5,245,512.61 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 264 years, 9 months from now.
Please review their action plans below:


**1) [visitskyworld](https://gitcoin.co/visitskyworld) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/shapeshift/web/949/100027615?mutate_worker_action=approve&worker=visitskyworld) | [reject worker](https://gitcoin.co/issue/shapeshift/web/949/100027615?mutate_worker_action=reject&worker=visitskyworld))_.**

`cypress-websocket-testing` and  `mock-socket ` are useful for mocking websocket.
But I think cypress also supports websocket mocking
We can do like this:
```
cy.visit('/', {
      onBeforeLoad: (win) => {
        cy.stub(win, 'WebSocket', (url) => {
          MockServer = new Server(url)
          MockServer.on('connection', (socket) => {
            socket.send(alert)
            setTimeout(() => {
              socket.close()
              MockServer.close()
            }, 1000)
          })
          return new WebSocket(url) /* mock socket  */
        })
      }
    })
```
Please tell me which method or library you want to mock websocket server.
Thanks.

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/shapeshift/web/949/100027615).

**Comment by 0xdef1cafe** [comment]
@visitskyworld how are you going with this one?

**Comment by visitskyworld** [comment]
As transaction type is different from ethereum socket server and mock server, mocked transactions are not rendered in dashboard page.

Real transaction type:
```
{
  address: '0xA44C286BA83Bb771cd0107B2c1Df678435Bd1535',
  blockHash: '0xc2f42f9dfbeb1c600ecd13dbcd624af1ea70631e13ad4d972b83a2a7c805360c',
  blockHeight: 1,
  blockTime: 1637201790,
  confirmations: 29807,
  caip2: 'eip155:1',
  txid: '0x5b25d67e43ba2cdfcb584c8069330874e838607d06b9dc64bf174547ca11e171',
  fee: {
    caip19: 'eip155:1/slip44:60',
    value: '3192000000000000'
  },
  status: chainAdapters.TxStatus.Confirmed,
  chain: ChainTypes.Ethereum,
  transfers: [
    {
      caip19: 'eip155:1/slip44:60',
      from: '0xA44C286BA83Bb771cd0107B2c1Df678435Bd1535',
      to: '0x67ffc0d460c38CeEe3a29A94def9Ff828E92E165',
      value: '230757160394687',
      type: chainAdapters.TxType.Send
    }
  ]
}
```
Mock transaction in `cypress/factories/cypress/transactions.ts`:
```
      {
        txid: '0x8e3528c933483770a3c8377c2ee7e34f846908653168188fd0d90a20b295d002',
        status: 'confirmed',
        from: '0xB3DD70991aF983Cf82d95c46C24979ee98348ffa',
        to: '0x642F4Bda144C63f6DC47EE0fDfbac0a193e2eDb7',
        blockHash: '0x94228c1b7052720846e2d7b9f36de30acf45d9a06ec483bd4433c5c38c8673a8',
        blockHeight: 12267105,
        confirmations: 1837143,
        timestamp: 1618788849,
        value: '737092621690531649',
        fee: '3180000000009000'
      },
```
Should I change mock data? @0xApotheosis

**Comment by 0xApotheosis** [comment]
> Should I change mock data? @0xApotheosis 

@visitskyworld absolutely - please update them to whether accurately represents the actual data received via WebSockets.

The current transaction file is not yet used anywhere, and is more a placeholder. Adapt it to your needs 👌

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. **Submitted** 4. Done 

<hr>

__Work for 2200.0 FOX  has been submitted by__:

 1. [@visitskyworld](https://gitcoin.co/visitskyworld)



@0xdef1cafe please take a look at the submitted work:
* [PR](https://github.com/shapeshift/web/pull/1013) by @visitskyworld

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/shapeshift/web/949/100027615)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/shapeshift/web/949/100027615).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://discord.gg/gitcoin/'>Gitcoin's Discord</a>
* $5,100,489.55 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by 0xApotheosis** [comment]
@visitskyworld how are you going with the requested changes on this one?

Are you able to provide an indication of when you'll be able to update the PR for review?

**Comment by visitskyworld** [comment]
@0xApotheosis , sorry for missing this one. I will push updates asap.
But regarding implementation, I need your detailed explanation how to make mock server extensible and add mechanism for websocket intercepting.
> As I understand it, the instance of web Cypress is talking to needs to point to the local, mocked, WebSocket server (ws://localhost:8080).
We can handle this by updating the CI configuration, though my concern is that this causes all WebSocket requests to be intercepted.
Can we implement a pass-through mechanism, where if we haven't mocked a subscription topic, we'll send that subscription request through to the actual server?
This would require the more generic WebSocket server logic noted above, and will prevent Cypress from breaking when we add new WebSocket topics that are not explicitly mocked.

**Comment by 0xApotheosis** [comment]
Hey @visitskyworld - I can see you've made some of the suggested changes in a new branch.

Will you be able to get the remaining changes to us in the next 24 hours?

If not, in the interest of expediency we may need to bring the remaining work in-house, or re-bounty what's left - we will of course issue a partial payment for the work done thus far!

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. **Submitted** 4. Done 

<hr>

__The funding of 2200.0 FOX  (603.39 USD @ $0.27/FOX) attached to this issue has been **cancelled** by the bounty submitter__

 
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://discord.gg/gitcoin/'>Gitcoin's Discord</a>
* $9,312,313.85 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by 0xean** [comment]
#1103

**Comment by 0xApotheosis** [comment]
@0xean I'm reopening this, as it's actually a related but separate issue to https://github.com/shapeshift/web/issues/1103.

**Comment by 0xApotheosis** [comment]
@GMSteuart is keen to bring this home 🤙

**Comment by GMSteuart** [comment]
@0xApotheosis The [cypress-websocket-testing](https://github.com/lensesio/cypress-websocket-testing) package only adds support for mocking websocket connections and not mocking websocket responses. While there has been some additional code added to try and do some simple response mocking in the existing branch, mock responses are not resulting in ui updates. Going to take a step back from the current implementation in the PR and get a better bird's eye view.

**Comment by 0xApotheosis** [comment]
Sounds good @GMSteuart - I'm not wedded to the current implementation in any way, and I'd even caution against taking too much inspiration from it.

I'd recommend taking a holistic look at how you'd approach it from scratch.

**Comment by 0xean** [comment]
@GMSteuart - can you provide an update on the status of this one? TY!

**Comment by 0xean** [comment]
@GMSteuart - still working on it, modifying scope to simplify with @0xApotheosis

**Comment by MBMaria** [comment]
@0xApotheosis do you have a rough estimate how many expected hours this ticket should take so I can update the amount the bounty needs to be in Dework?

**Comment by 0xApotheosis** [comment]
It depends on the implementation method chosen, as there are a few possible approaches. I'd expect 2-3 days of effort - @GMSteuart may have further thoughts, as I believe he is (or was?) working on this.

**Comment by GMSteuart** [comment]
@0xApotheosis I was waiting on refactoring to be completed with lib/chain-adapters package before starting since mocking of web sockets is dependent on it's implementation. Ideally, there would be some reusable code for both testing the chain-adapters package and mocking the web sockets.

@MBMaria I would agree with the time estimate @0xApotheosis gave of 2-3 days of effort.

**Comment by MBMaria** [comment]
Thank you both. I’ll update the bounty in Dework!

**Comment by 0xApotheosis** [comment]
We should kill this - the main reason we wanted it is to get TX history via websockets.

Recent architectural changes, i.e. https://github.com/shapeshift/web/pull/1900, mean that we now get TX history RESTfully, and so this is no longer important.

**Comment by 0xApotheosis** [comment]
@MBMaria FYI ^
